### ISSUE TITLE
grpc-js: Uncaught Exception

### ISSUE THREAD

**Issue by artificial-aidan** [grpc/grpc-node#1871]
### Problem description
When heavily loaded, grpc-js can throw an uncaught exception in the `sendMetadata` call if the client closes the stream before metadata is sent.

Similar to https://github.com/grpc/grpc-node/issues/984 but it seems that the `checkCancelled` is not enough. When I add a try/catch block around the `respond` call [here](https://github.com/grpc/grpc-node/blob/92cc956dec0cb933f3b77fb93c23271f4fa3b17b/packages/grpc-js/src/server-call.ts#L453) then the crash goes away.

I added some logging, and the `checkCancelled` call succeeds, but then the stream is destroyed before the `respond` call is made.

I also added a check for `closed` in `checkCancelled` but that didn't help.

Stack Trace:

```
2021-08-04T19:46:46.426259133Z node:internal/errors:464
2021-08-04T19:46:46.426304830Z     ErrorCaptureStackTrace(err);
2021-08-04T19:46:46.426309525Z     ^
2021-08-04T19:46:46.426311929Z 
2021-08-04T19:46:46.426314239Z Error [ERR_HTTP2_INVALID_STREAM]: The stream has been destroyed
2021-08-04T19:46:46.426316773Z     at new NodeError (node:internal/errors:371:5)
2021-08-04T19:46:46.426319997Z     at ServerHttp2Stream.respond (node:internal/http2/core:2718:13)
2021-08-04T19:46:46.426323573Z     at Http2ServerCallStream.sendMetadata (/bin/node_modules/@grpc/grpc-js/build/src/server-call.js:265:21)
2021-08-04T19:46:46.426327148Z     at Http2ServerCallStream.sendStatus (/bin/node_modules/@grpc/grpc-js/build/src/server-call.js:382:18)
2021-08-04T19:46:46.426330524Z     at Http2ServerCallStream.sendError (/bin/node_modules/@grpc/grpc-js/build/src/server-call.js:405:14)
2021-08-04T19:46:46.426333963Z     at ServerHttp2Stream.<anonymous> (/bin/node_modules/@grpc/grpc-js/build/src/server-call.js:317:26)
2021-08-04T19:46:46.426337894Z     at Object.onceWrapper (node:events:513:28)
2021-08-04T19:46:46.426341397Z     at ServerHttp2Stream.emit (node:events:406:35)
2021-08-04T19:46:46.426345359Z     at ServerHttp2Stream.emit (node:domain:475:12)
2021-08-04T19:46:46.426347948Z     at endReadableNT (node:internal/streams/readable:1331:12) {
```

### Reproduction steps
Heavily load a GRPC server, with a client that has a timeout, and it will crash.

This is my first foray into `grpc-js` development, if you think my explanation is a possibility I could spend some more time trying to create a minimally reproducible case.

### Environment
 - Node 16.1 Alpine
 - Node version 15.8/16.1
 - Node installation method: node docker image
 - grpc-js@1.3.36

**Comment by murgatroid99** [comment]
> I added some logging, and the `checkCancelled` call succeeds, but then the stream is destroyed before the `respond` call is made.

What exactly did your logging show? Are you saying that the `checkCancelled` call in `sendMetadata` returned `false`, but on a later line, before the call to `respond`, `this.stream.destroyed` was `true`?

**Comment by artificial-aidan** [comment]
I will run a test with that.

I had trace turned on, and I saw a log message that the stream was closed, but I did not check the `this.stream.destroyed`, give me a few.

**Comment by murgatroid99** [comment]
That makes it even more surprising that checking `this.stream.closed` in `isCancelled` did not resolve the problem.

**Comment by artificial-aidan** [comment]
> That makes it even more surprising that checking `this.stream.closed` in `isCancelled` did not resolve the problem.

I know. I thought for sure I had found the problem there.

**Comment by artificial-aidan** [comment]
For this code:
```
  private checkCancelled(): boolean {
    // console.log('hello')
    // trace('checkCancelled')
    /* In some cases the stream can become destroyed before the close event
     * fires. That creates a race condition that this check works around */
    // if (this.stream.destroyed || this.stream.closed ) {
    if (this.stream.destroyed) {
      this.cancelled = true;
    }
    if (this.stream.closed) {
      console.log('checkCancelled ERROR stream was closed but not destroyed')
    }
    return this.cancelled;
  }

  sendMetadata(customMetadata?: Metadata) {
    // trace('sendMetadata')
    if (this.checkCancelled()) {
      return;
    }

    if (this.metadataSent) {
      return;
    }

    this.metadataSent = true;
    const custom = customMetadata ? customMetadata.toHttp2Headers() : null;
    // TODO(cjihrig): Include compression headers.
    const headers = Object.assign({}, defaultResponseHeaders, custom);
    if (this.stream.destroyed) {
      console.log('ERRROOOORRR SHOULD NOT BE HERE: destroyed')
    }
    if (this.stream.closed) {
      console.log('ERRROOOORRR SHOULD NOT BE HERE: closed')
    }
    try {
      this.stream.respond(headers, defaultResponseOptions);
    } catch(err) {
      console.log(`Error responding: ${err}`)
      console.log(`Stream was destroyed: ${this.stream.destroyed}. Stream was closed: ${this.stream.closed}`)
    }
  }
```

I get this log:

```
checkCancelled ERROR stream was closed but not destroyed
checkCancelled ERROR stream was closed but not destroyed
checkCancelled ERROR stream was closed but not destroyed
ERRROOOORRR SHOULD NOT BE HERE: closed
Error responding: Error [ERR_HTTP2_INVALID_STREAM]: The stream has been destroyed
Stream was destroyed: false. Stream was closed: true
```

That would hint that adding the check for closed would do it, but it didn't do it yesterday. Testing again now.

**Comment by artificial-aidan** [comment]
I must have gotten caught up in my testing yesterday.

Adding a check for `this.stream.closed` has solved the problem. I'll run a bit more stress on it, but it looks good.

**Comment by murgatroid99** [comment]
OK, great. I can get a fix for that out soon, probably tomorrow.

**Comment by murgatroid99** [comment]
OK, the added check for `this.stream.closed` is now out in version 1.3.7. Please try it out.
